File: example/android/app/src/main/kotlin/net/cubiclab/clipboard/ClipboardPlugin.kt
package net.cubiclab.clipboard

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Build
import android.util.Base64
import android.content.ContentResolver
import android.provider.MediaStore
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class ClipboardPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var context: Context? = null
    private var clipboardManager: ClipboardManager? = null
    private var clipboardChangeListener: ClipboardManager.OnPrimaryClipChangedListener? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        clipboardManager = context?.getSystemService(Context.CLIPBOARD_SERVICE) as? ClipboardManager

        methodChannel = MethodChannel(binding.binaryMessenger, "net.cubiclab.clipboard/methods")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "net.cubiclab.clipboard/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        stopMonitoring()
        context = null
        clipboardManager = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "copy" -> {
                val text = call.argument<String>("text") ?: ""
                if (text.isEmpty()) {
                    result.error("EMPTY_TEXT", "Text cannot be empty", null)
                    return
                }
                try {
                    val clip = ClipData.newPlainText("text", text)
                    clipboardManager?.setPrimaryClip(clip)
                    result.success(true)
                } catch (e: Exception) {
                    result.error("COPY_ERROR", e.message, null)
                }
            }
            "copyRichText" -> {
                val text = call.argument<String>("text") ?: ""
                val html = call.argument<String>("html")
                if (text.isEmpty() && (html == null || html.isEmpty())) {
                    result.error("EMPTY_CONTENT", "Either text or html must be provided", null)
                    return
                }
                try {
                    if (html != null && html.isNotEmpty()) {
                        val clip = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                            ClipData.newHtmlText("html", text, html)
                        } else {
                            ClipData.newPlainText("text", text)
                        }
                        clipboardManager?.setPrimaryClip(clip)
                    } else {
                        val clip = ClipData.newPlainText("text", text)
                        clipboardManager?.setPrimaryClip(clip)
                    }
                    result.success(true)
                } catch (e: Exception) {
                    result.error("COPY_RICH_ERROR", e.message, null)
                }
            }
            "copyMultiple" -> {
                val formats = call.argument<Map<String, Any>>("formats") ?: emptyMap()
                if (formats.isEmpty()) {
                    result.error("EMPTY_FORMATS", "At least one format must be provided", null)
                    return
                }
                try {
                    val text = formats["text/plain"]?.toString() ?: ""
                    val html = formats["text/html"]?.toString()
                    val imageBytes = formats["image/png"] as? List<Int>

                    // Handle image first (highest priority)
                    if (imageBytes != null && imageBytes.isNotEmpty()) {
                        val byteArray = imageBytes.map { it.toByte() }.toByteArray()
                        val bitmap = bytesToBitmap(byteArray)
                        if (bitmap != null) {
                            val imageUri = saveBitmapToCache(bitmap)
                            if (imageUri != null) {
                                val clip = ClipData.newUri(context?.contentResolver, "image", imageUri)
                                if (text.isNotEmpty()) {
                                    clip.addItem(ClipData.Item(text))
                                }
                                clipboardManager?.setPrimaryClip(clip)
                                result.success(true)
                                return
                            }
                        }
                    }

                    // Fallback to HTML or text
                    if (html != null && html.isNotEmpty() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                        val clip = ClipData.newHtmlText("html", text, html)
                        clipboardManager?.setPrimaryClip(clip)
                    } else if (text.isNotEmpty()) {
                        val clip = ClipData.newPlainText("text", text)
                        clipboardManager?.setPrimaryClip(clip)
                    }
                    result.success(true)
                } catch (e: Exception) {
                    result.error("COPY_MULTIPLE_ERROR", e.message, null)
                }
            }
            "copyImage" -> {
                val imageBytes = call.argument<List<Int>>("imageBytes")
                if (imageBytes == null || imageBytes.isEmpty()) {
                    result.error("EMPTY_IMAGE", "Image bytes cannot be empty", null)
                    return
                }
                try {
                    val byteArray = imageBytes.map { it.toByte() }.toByteArray()
                    val bitmap = bytesToBitmap(byteArray)
                    if (bitmap == null) {
                        result.error("INVALID_IMAGE", "Failed to decode image", null)
                        return
                    }
                    val imageUri = saveBitmapToCache(bitmap)
                    if (imageUri == null) {
                        result.error("SAVE_ERROR", "Failed to save image to cache", null)
                        return
                    }
                    val clip = ClipData.newUri(context?.contentResolver, "image", imageUri)
                    clipboardManager?.setPrimaryClip(clip)
                    result.success(true)
                } catch (e: Exception) {
                    result.error("COPY_IMAGE_ERROR", e.message, null)
                }
            }
            "paste" -> {
                try {
                    val clipData = clipboardManager?.primaryClip
                    val text = clipData?.getItemAt(0)?.text?.toString() ?: ""
                    result.success(mapOf("text" to text))
                } catch (e: Exception) {
                    result.error("PASTE_ERROR", e.message, null)
                }
            }
            "pasteRichText" -> {
                try {
                    val clipData = clipboardManager?.primaryClip
                    val item = clipData?.getItemAt(0)
                    val text = item?.text?.toString() ?: ""
                    val html = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                        item?.htmlText
                    } else {
                        null
                    }
                    
                    // Try to get image from clipboard
                    val imageBytes = getImageFromClipboard(item)
                    
                    val resultMap = mutableMapOf<String, Any?>(
                        "text" to text,
                        "html" to html,
                        "imageBytes" to imageBytes,
                        "timestamp" to System.currentTimeMillis()
                    )
                    result.success(resultMap)
                } catch (e: Exception) {
                    result.error("PASTE_RICH_ERROR", e.message, null)
                }
            }
            "pasteImage" -> {
                try {
                    val clipData = clipboardManager?.primaryClip
                    val item = clipData?.getItemAt(0)
                    val imageBytes = getImageFromClipboard(item)
                    if (imageBytes != null) {
                        result.success(mapOf("imageBytes" to imageBytes))
                    } else {
                        result.success(mapOf("imageBytes" to null))
                    }
                } catch (e: Exception) {
                    result.error("PASTE_IMAGE_ERROR", e.message, null)
                }
            }
            "getContentType" -> {
                try {
                    val clipData = clipboardManager?.primaryClip
                    if (clipData == null || clipData.itemCount == 0) {
                        result.success("empty")
                        return
                    }
                    val item = clipData.getItemAt(0)
                    val hasText = item?.text != null
                    val hasHtml = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                        item?.htmlText != null
                    } else {
                        false
                    }
                    val hasImage = item?.uri != null && isImageUri(item.uri)
                    
                    when {
                        hasImage && (hasText || hasHtml) -> result.success("mixed")
                        hasImage -> result.success("image")
                        hasText && hasHtml -> result.success("mixed")
                        hasHtml -> result.success("html")
                        hasText -> result.success("text")
                        else -> result.success("empty")
                    }
                } catch (e: Exception) {
                    result.success("unknown")
                }
            }
            "hasData" -> {
                try {
                    val clipData = clipboardManager?.primaryClip
                    val hasData = clipData != null && clipData.itemCount > 0
                    result.success(hasData)
                } catch (e: Exception) {
                    result.success(false)
                }
            }
            "clear" -> {
                try {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                        clipboardManager?.clearPrimaryClip()
                    } else {
                        val clip = ClipData.newPlainText("", "")
                        clipboardManager?.setPrimaryClip(clip)
                    }
                    result.success(true)
                } catch (e: Exception) {
                    result.error("CLEAR_ERROR", e.message, null)
                }
            }
            "getDataSize" -> {
                try {
                    val clipData = clipboardManager?.primaryClip
                    val text = clipData?.getItemAt(0)?.text?.toString() ?: ""
                    result.success(text.length)
                } catch (e: Exception) {
                    result.success(0)
                }
            }
            "startMonitoring" -> {
                startMonitoring()
                result.success(true)
            }
            "stopMonitoring" -> {
                stopMonitoring()
                result.success(true)
            }
            else -> result.notImplemented()
        }
    }

    private fun startMonitoring() {
        if (clipboardChangeListener != null) {
            return
        }
        clipboardChangeListener = ClipboardManager.OnPrimaryClipChangedListener {
            try {
                val clipData = clipboardManager?.primaryClip
                val item = clipData?.getItemAt(0)
                val text = item?.text?.toString() ?: ""
                val html = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    item?.htmlText
                } else {
                    null
                }
                val eventMap = mapOf(
                    "text" to text,
                    "html" to html,
                    "timestamp" to System.currentTimeMillis()
                )
                eventSink?.success(eventMap)
            } catch (e: Exception) {
                // Ignore errors
            }
        }
        clipboardManager?.addPrimaryClipChangedListener(clipboardChangeListener!!)
    }

    private fun stopMonitoring() {
        clipboardChangeListener?.let {
            clipboardManager?.removePrimaryClipChangedListener(it)
            clipboardChangeListener = null
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        startMonitoring()
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        stopMonitoring()
    }

    private fun bytesToBitmap(bytes: ByteArray): Bitmap? {
        return try {
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        } catch (e: Exception) {
            null
        }
    }

    private fun saveBitmapToCache(bitmap: Bitmap): Uri? {
        return try {
            val context = this.context ?: return null
            val cacheDir = context.cacheDir
            val imageFile = File(cacheDir, "clipboard_image_${System.currentTimeMillis()}.png")
            
            FileOutputStream(imageFile).use { out ->
                bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
            }
            
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                android.content.FileProvider.getUriForFile(
                    context,
                    "${context.packageName}.fileprovider",
                    imageFile
                )
            } else {
                Uri.fromFile(imageFile)
            }
        } catch (e: Exception) {
            null
        }
    }

    private fun getImageFromClipboard(item: ClipData.Item?): List<Int>? {
        return try {
            val uri = item?.uri ?: return null
            if (!isImageUri(uri)) return null
            
            val context = this.context ?: return null
            val inputStream = context.contentResolver.openInputStream(uri) ?: return null
            
            val bitmap = BitmapFactory.decodeStream(inputStream)
            inputStream.close()
            
            if (bitmap == null) return null
            
            val outputStream = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream)
            val byteArray = outputStream.toByteArray()
            outputStream.close()
            
            byteArray.toList()
        } catch (e: Exception) {
            null
        }
    }

    private fun isImageUri(uri: Uri): Boolean {
        val mimeType = context?.contentResolver?.getType(uri)
        return mimeType?.startsWith("image/") == true
    }
}

File: ios/Classes/ClipboardPlugin.swift
import Flutter
import UIKit

public class ClipboardPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var clipboardChangeObserver: NSObjectProtocol?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "net.cubiclab.clipboard/methods",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "net.cubiclab.clipboard/events",
            binaryMessenger: registrar.messenger()
        )
        
        let instance = ClipboardPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "copy":
            guard let args = call.arguments as? [String: Any],
                  let text = args["text"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Text is required", details: nil))
                return
            }
            if text.isEmpty {
                result(FlutterError(code: "EMPTY_TEXT", message: "Text cannot be empty", details: nil))
                return
            }
            UIPasteboard.general.string = text
            result(true)
            
        case "copyRichText":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid arguments", details: nil))
                return
            }
            let text = args["text"] as? String ?? ""
            let html = args["html"] as? String
            
            if text.isEmpty && (html == nil || html!.isEmpty) {
                result(FlutterError(code: "EMPTY_CONTENT", message: "Either text or html must be provided", details: nil))
                return
            }
            
            if let html = html, !html.isEmpty {
                UIPasteboard.general.setValue(html, forPasteboardType: "public.html")
                if !text.isEmpty {
                    UIPasteboard.general.string = text
                }
            } else {
                UIPasteboard.general.string = text
            }
            result(true)
            
        case "copyMultiple":
            guard let args = call.arguments as? [String: Any],
                  let formats = args["formats"] as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid arguments", details: nil))
                return
            }
            if formats.isEmpty {
                result(FlutterError(code: "EMPTY_FORMATS", message: "At least one format must be provided", details: nil))
                return
            }
            
            // Handle image first (highest priority)
            if let imageBytes = formats["image/png"] as? [Int], !imageBytes.isEmpty {
                let data = Data(imageBytes.map { UInt8($0 & 0xFF) })
                if let image = UIImage(data: data) {
                    UIPasteboard.general.image = image
                    if let text = formats["text/plain"] as? String, !text.isEmpty {
                        UIPasteboard.general.string = text
                    }
                    result(true)
                    return
                }
            }
            
            // Fallback to HTML or text
            if let text = formats["text/plain"] as? String {
                UIPasteboard.general.string = text
            }
            if let html = formats["text/html"] as? String {
                UIPasteboard.general.setValue(html, forPasteboardType: "public.html")
            }
            result(true)
            
        case "copyImage":
            guard let args = call.arguments as? [String: Any],
                  let imageBytes = args["imageBytes"] as? [Int] else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Image bytes are required", details: nil))
                return
            }
            if imageBytes.isEmpty {
                result(FlutterError(code: "EMPTY_IMAGE", message: "Image bytes cannot be empty", details: nil))
                return
            }
            let data = Data(imageBytes.map { UInt8($0 & 0xFF) })
            guard let image = UIImage(data: data) else {
                result(FlutterError(code: "INVALID_IMAGE", message: "Failed to decode image", details: nil))
                return
            }
            UIPasteboard.general.image = image
            result(true)
            
        case "paste":
            let text = UIPasteboard.general.string ?? ""
            result(["text": text])
            
        case "pasteRichText":
            let text = UIPasteboard.general.string ?? ""
            let html = UIPasteboard.general.value(forPasteboardType: "public.html") as? String
            let imageBytes = getImageBytesFromClipboard()
            result([
                "text": text,
                "html": html ?? NSNull(),
                "imageBytes": imageBytes ?? NSNull(),
                "timestamp": Int64(Date().timeIntervalSince1970 * 1000)
            ])
            
        case "pasteImage":
            let imageBytes = getImageBytesFromClipboard()
            if let bytes = imageBytes {
                result(["imageBytes": bytes])
            } else {
                result(["imageBytes": NSNull()])
            }
            
        case "getContentType":
            let text = UIPasteboard.general.string ?? ""
            let html = UIPasteboard.general.value(forPasteboardType: "public.html") as? String
            let hasImage = UIPasteboard.general.image != nil
            
            if hasImage && (!text.isEmpty || (html != nil && !html!.isEmpty)) {
                result("mixed")
            } else if hasImage {
                result("image")
            } else if text.isEmpty && (html == nil || html!.isEmpty) {
                result("empty")
            } else if !text.isEmpty && html != nil && !html!.isEmpty {
                result("mixed")
            } else if html != nil && !html!.isEmpty {
                result("html")
            } else {
                result("text")
            }
            
        case "hasData":
            let text = UIPasteboard.general.string ?? ""
            result(!text.isEmpty)
            
        case "clear":
            UIPasteboard.general.string = ""
            result(true)
            
        case "getDataSize":
            let text = UIPasteboard.general.string ?? ""
            result(text.count)
            
        case "startMonitoring":
            startMonitoring()
            result(true)
            
        case "stopMonitoring":
            stopMonitoring()
            result(true)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func startMonitoring() {
        if clipboardChangeObserver != nil {
            return
        }
        
        // iOS doesn't have native clipboard change notifications
        // We'll use a timer-based approach as fallback
        // The Dart side will handle the actual polling
        NotificationCenter.default.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.checkClipboardChange()
        }
    }
    
    private func stopMonitoring() {
        if let observer = clipboardChangeObserver {
            NotificationCenter.default.removeObserver(observer)
            clipboardChangeObserver = nil
        }
    }
    
    private func checkClipboardChange() {
        let text = UIPasteboard.general.string ?? ""
        let html = UIPasteboard.general.value(forPasteboardType: "public.html") as? String
        
        eventSink?([
            "text": text,
            "html": html ?? NSNull(),
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000)
        ])
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        startMonitoring()
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        stopMonitoring()
        return nil
    }
    
    private func getImageBytesFromClipboard() -> [Int]? {
        guard let image = UIPasteboard.general.image else {
            return nil
        }
        guard let imageData = image.pngData() else {
            return nil
        }
        return Array(imageData.map { Int($0) })
    }
    
    deinit {
        stopMonitoring()
    }
}

